100of100Service/src/test/java/dev/coms4156/project/HrDatabaseFacadeTest.java

```diff
@@ -1,29 +1,396 @@
  */
 public class HrDatabaseFacadeTest {
 
-  @BeforeAll
-  public static void setup() {
-    DatabaseConnection realConnection = MysqlConnection.getInstance();
-    HrDatabaseFacade.setConnection(realConnection);
+  private DatabaseConnection dbConnection;
+  private HrDatabaseFacade facade;
+  private int testOrganizationId = 1;
+
+  /**
+   * Sets up the test environment by initializing the DatabaseConnection
+   * and HrDatabaseFacade instance.
+   */
+  @BeforeEach
+  public void setup() {
+    dbConnection = InmemConnection.getInstance();
+    HrDatabaseFacade.setConnection(dbConnection);
+    facade = HrDatabaseFacade.getInstance(testOrganizationId);
+  }
+
+  @Test
+  @Order(1)
+  public void testGetEmployee_employeeNotInCache() throws Exception {
+    Field employeesField = HrDatabaseFacade.class.getDeclaredField("employees");
+    employeesField.setAccessible(true);
+    employeesField.set(facade, new ArrayList<>());
+
+    Employee employee = facade.getEmployee(1);
+    assertNotNull(employee, "Employee should be fetched from database when not in cache");
+
+    // Verify cache is updated
+    List<Employee> employeesCache = (List<Employee>) employeesField.get(facade);
+    assertFalse(employeesCache.isEmpty(), "Employees cache should be updated");
+    assertTrue(employeesCache.contains(employee),
+        "Employees cache should contain the fetched employee");
+  }
+
+  @Test
+  @Order(2)
+  public void testGetDepartment_departmentNotInCache() throws Exception {
+    Field departmentsField = HrDatabaseFacade.class.getDeclaredField("departments");
+    departmentsField.setAccessible(true);
+    departmentsField.set(facade, new ArrayList<>());
+
+    Department department = facade.getDepartment(1);
+    assertNotNull(department, "Department should be fetched from database when not in cache");
+  }
+
+  @Test
+  @Order(3)
+  public void testRemoveEmployeeFromDepartment_DepartmentWithMultipleEmployees() {
+    // Add multiple employees to a department
+    Employee employee1 = new Employee(0, "Employee1", new Date());
+    Employee employee2 = new Employee(0, "Employee2", new Date());
+    Employee employee3 = new Employee(0, "Employee3", new Date());
+
+    Employee addedEmployee1 = facade.addEmployeeToDepartment(1, employee1);
+    Employee addedEmployee2 = facade.addEmployeeToDepartment(1, employee2);
+    Employee addedEmployee3 = facade.addEmployeeToDepartment(1, employee3);
+
+    boolean removed = facade.removeEmployeeFromDepartment(1, addedEmployee2.getId());
+    assertTrue(removed, "Employee should be removed successfully");
+
+    facade.removeEmployeeFromDepartment(1, addedEmployee1.getId());
+    facade.removeEmployeeFromDepartment(1, addedEmployee3.getId());
+  }
+
+  @Test
+  @Order(4)
+  public void testRemoveDepartment_Successful() throws Exception {
+    // Create a mock DatabaseConnection that returns true for removeDepartment
+    DatabaseConnection mockDbConnection = new InmemConnection() {
```

Comment:
This line causes the first compilation error of `Error:  HrDatabaseFacadeTest.java:[91,47] cannot inherit from final dev.coms4156.project.InmemConnection`, because currently in our Main Branch, PMD required us to make those classes with only private constructor (before your change, it only has one private constructor) to be final class, who cannot be inherited by others. 

So, your branch itself is okay to compile, but when GitHub trying to merge your branch with our main branch (btw, it is strange it did not find conflict), this issue occurs. And causes the CI failed.

Previously, I was thinking to change the production file to satisfy the test purpose may not be a good idea, but given that one of the purpose of `InmemConnection` is to support testing, I agree with your approach.

As for suggestion, please try to pull the main branch into your branch locally, so that you can remove the `final` qualifier and push again.